### ISSUE TITLE
Dolt panics on subquery in IF statement in procedure

### ISSUE THREAD

**Issue by arvidfm** [dolthub/dolt#7944]
A procedure like the following causes Dolt to panic:
```sql
CREATE PROCEDURE populate_if_empty()
BEGIN
    IF (SELECT COUNT(*) FROM my_table) = 0 THEN
        INSERT INTO my_table (column1, column2) VALUES (1, 2);
    END IF;
END
```

MWE, works with Percona Server 8.3.0:
```sql
DROP PROCEDURE IF EXISTS populate_if_empty;

DELIMITER //

CREATE PROCEDURE populate_if_empty()
BEGIN
    IF (SELECT 0) = 0 THEN
        SELECT 'hi';
    END IF;
END //

DELIMITER ;

CALL populate_if_empty();

DROP PROCEDURE populate_if_empty;
```

<details>
<summary>Full stack trace</summary>

```
runtime error: invalid memory address or nil pointer dereference
go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.linux-amd64/src/runtime/panic.go:261 (0x4470fd)
        panicmem: panic(memoryError)
go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.linux-amd64/src/runtime/signal_unix.go:881 (0x461ce4)
        sigpanic: panicmem()
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/plan/subquery.go:366 (0x2462afd)
        com/dolthub/go-mysql-server/sql/plan.(*Subquery).evalMultiple: iter, err := s.b.Build(ctx, q, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/plan/subquery.go:175 (0x246069d)
        com/dolthub/go-mysql-server/sql/plan.(*Subquery).Eval: rows, err := s.evalMultiple(ctx, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/expression/comparison.go:156 (0x19352fe)
        com/dolthub/go-mysql-server/sql/expression.(*comparison).evalLeftAndRight: left, err := c.Left().Eval(ctx, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/expression/comparison.go:93 (0x193439b)
        com/dolthub/go-mysql-server/sql/expression.(*comparison).Compare: left, right, err := c.evalLeftAndRight(ctx, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/expression/comparison.go:301 (0x19374ad)
        com/dolthub/go-mysql-server/sql/expression.(*Equals).Eval: result, err := e.Compare(ctx, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/proc.go:77 (0x295793a)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildIfElseBlock: condition, err := ifConditional.Condition.Eval(ctx, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:344 (0x29359e4)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExecNoAnalyze: return b.buildIfElseBlock(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:27 (0x293500f)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExec: iter, err := b.buildNodeExecNoAnalyze(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/other.go:234 (0x2949204)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildBlock.func1: subIter, err := b.buildNodeExec(ctx, s, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/other.go:274 (0x2948e28)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildBlock: }()
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:210 (0x293f838)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExecNoAnalyze: return b.buildBlock(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:27 (0x293500f)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExec: iter, err := b.buildNodeExecNoAnalyze(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/proc.go:139 (0x2958456)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildBeginEndBlock: rowIter, err := b.buildNodeExec(ctx, n.Block, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:110 (0x293fb04)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExecNoAnalyze: return b.buildBeginEndBlock(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:27 (0x293500f)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExec: iter, err := b.buildNodeExecNoAnalyze(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/rel.go:331 (0x2967a73)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildProcedure: return b.buildNodeExec(ctx, n.Body, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:240 (0x293b944)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExecNoAnalyze: return b.buildProcedure(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:27 (0x293500f)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExec: iter, err := b.buildNodeExecNoAnalyze(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/proc.go:228 (0x2959c14)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildCall: innerIter, err := b.buildNodeExec(ctx, n.Procedure, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:98 (0x29384a4)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExecNoAnalyze: return b.buildCall(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:27 (0x293500f)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExec: iter, err := b.buildNodeExecNoAnalyze(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/builder.go:33 (0x28df9e6)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).Build: return b.buildNodeExec(ctx, n, r)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/transaction.go:305 (0x299aa84)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildTransactionCommittingNode: iter, err := b.Build(ctx, n.Child(), row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:48 (0x29360fd)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExecNoAnalyze: return b.buildTransactionCommittingNode(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:27 (0x293500f)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExec: iter, err := b.buildNodeExecNoAnalyze(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/builder.go:33 (0x28df9e6)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).Build: return b.buildNodeExec(ctx, n, r)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/other.go:326 (0x294a3e4)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildQueryProcess: iter, err := b.Build(ctx, n.Child(), row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:58 (0x29449a4)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExecNoAnalyze: return b.buildQueryProcess(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/node_builder.gen.go:27 (0x293500f)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).buildNodeExec: iter, err := b.buildNodeExecNoAnalyze(ctx, n, row)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/sql/rowexec/builder.go:33 (0x28df9e6)
        com/dolthub/go-mysql-server/sql/rowexec.(*BaseBuilder).Build: return b.buildNodeExec(ctx, n, r)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/engine.go:408 (0x2b7082a)
        com/dolthub/go-mysql-server.(*Engine).QueryWithBindings: iter, err := e.Analyzer.ExecBuilder.Build(ctx, analyzed, nil)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/server/handler.go:961 (0x2f32c6d)
        com/dolthub/go-mysql-server/server.(*Handler).executeQuery: return h.e.QueryWithBindings(ctx, query, parsed, bindings)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/server/handler.go:409 (0x2f2c66e)
        com/dolthub/go-mysql-server/server.(*Handler).doQuery: schema, rowIter, err := queryExec(ctx, query, parsed, analyzedPlan, bindings)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/server/handler.go:665 (0x2f2fe66)
        com/dolthub/go-mysql-server/server.(*Handler).errorWrappedDoQuery: remainder, err := h.doQuery(c, query, parsed, nil, mode, h.executeQuery, bindings, callback)
go/pkg/mod/github.com/dolthub/go-mysql-server@v0.18.2-0.20240529234304-47f89331b0ea/server/handler.go:283 (0x2f2b214)
        com/dolthub/go-mysql-server/server.(*Handler).ComMultiQuery: return h.errorWrappedDoQuery(c, query, nil, MultiStmtModeOn, nil, callback)
go/pkg/mod/github.com/dolthub/vitess@v0.0.0-20240524204318-4c419c3de9c4/go/mysql/conn.go:1548 (0xd02781)
        com/dolthub/vitess/go/mysql.(*Conn).execQuery: remainder, err = handler.ComMultiQuery(c, query, resultsCB)
go/pkg/mod/github.com/dolthub/vitess@v0.0.0-20240524204318-4c419c3de9c4/go/mysql/conn.go:991 (0xcfa5f8)
        com/dolthub/vitess/go/mysql.(*Conn).handleNextCommand: for query, err = c.execQuery(query, handler, multiStatements); err == nil && query != ""; {
go/pkg/mod/github.com/dolthub/vitess@v0.0.0-20240524204318-4c419c3de9c4/go/mysql/server.go:535 (0xd35610)
        com/dolthub/vitess/go/mysql.(*Listener).handle: err := c.handleNextCommand(l.handler)
go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.3.linux-amd64/src/runtime/asm_amd64.s:1695 (0x4845a0)
        goexit: BYTE    $0x90   // NOP
```
</details>

**Comment by fulghum** [comment]
Hi @arvidfm, thanks for the excellent repro case. I'll debug through this today and see what's going on.

**Comment by fulghum** [comment]
This was a really good find. Thank you for taking the time to report it and providing such excellent repro steps. 🙏 The problem was that when executing the `CALL` statement, we plug in the stored procedure logic and we weren't analyzing the subquery expression in the `IF` statement correctly. I fixed this case, added a new test for it, and got it merged in. We'll get a release out later today with the fix. 

We're also discussing some similar cases that likely need other tweaks to run correctly, such as when the if conditional needs to access tables/columns/functions/etc from the catalog. We need to figure out how the row data gets passed around; we'll keep discussing how to best handle that.

**Comment by fulghum** [comment]
@arvidfm – The fix for this issue has been released in [Dolt 1.39.3](https://github.com/dolthub/dolt/releases/tag/v1.39.3). Let us know if you hit any other snags! 

We suspect there are still some more cases with subqueries in `IF` conditionals that need a little more work. We're discussing the best way to structure that and will keep the ball moving, but don't hesitate to let us know if you hit another case that doesn't work for you and we can probably fast-track a fix for it.
